File: src/com/googlecode/yadic/SimpleContainer.scala
package com.googlecode.yadic

import java.lang.Class
import java.util.HashMap
import java.util.concurrent.Callable

class SimpleContainer(missingHandler: (Class[_]) => Object) extends Container {
  def this() = this ((aClass: Class[_]) => {throw new ContainerException(aClass.getName + " not found in container")})

  def this(resolver: Resolver) = this ((aCLass: Class[_]) => resolver.resolve(aCLass))

  val activators = new HashMap[Class[_], Callable[_]]

  def resolve(aClass: Class[_]): Object =
    activators.get(aClass) match {
      case null => missingHandler(aClass)
      case activator: Callable[_] => activator.call().asInstanceOf[Object]
    }

  def get[Type](aClass: Class[Type]): Type = resolve(aClass).asInstanceOf[Type]

  def getActivator[Type](aClass: Class[Type]): Callable[Type] = activators.get(aClass).asInstanceOf[Callable[Type]]

  def add[Concrete](concrete: Class[Concrete]): Container = add(concrete, () => create(concrete))

  def add[Interface, Concrete <: Interface](interface: Class[Interface], concrete: Class[Concrete]): Container = add(interface, () => create(concrete))

  def addInstance(instance: Object): Container = add(instance.getClass.asInstanceOf[Class[Object]], () => instance)

  def addInstance[Interface, Concrete <: Interface](anInterface: Class[Interface], instance: Concrete) = add(anInterface, () => instance)

  def addActivator[Type, ExtendsType <: Type](aClass: Class[Type], activator: Callable[ExtendsType]): Container = add(aClass, () => activator.call())

  def addActivator[Type, AnActivator <: Callable[Type]](aClass: Class[Type], activator: Class[AnActivator]): Container = add(activator).add(aClass, () => get(activator).call)

  def add[Type](aClass: Class[Type], activator: () => Type): Container = {
    activators.containsKey(aClass) match {
      case true => throw new ContainerException(aClass.getName + " already added to container")
      case false => activators.put(aClass, new LazyActivator[Type](activator))
    }
    this
  }

  def decorate[I, C <: I](interface: Class[I], concrete: Class[C]): Container = {
    val existing = activators.get(interface)
    activators.put(interface, new LazyActivator[I](() => create(concrete, (aClass: Class[_]) => {
      if (aClass.equals(interface)) existing.call() else get(aClass)
    })))
    this
  }

  def remove[T](aClass: Class[T]): Callable[T] = activators.remove(aClass).asInstanceOf[Callable[T]]

  def contains[T](aClass: Class[T]): Boolean = activators.containsKey(aClass)

  def create[C](concrete: Class[C]): C = create(concrete, resolve)

  def create[C](concrete: Class[C], resolver: (Class[_]) => Any): C = {
    val constructors = concrete.getConstructors.toList.sort(_.getParameterTypes.length > _.getParameterTypes.length)
    val exceptions = constructors.map(constructor => {
      try {
        val instances = constructor.getParameterTypes.map(resolver(_).asInstanceOf[Object])
        return constructor.newInstance(instances: _*).asInstanceOf[C]
      } catch {
        case e:ContainerException => e
      }
    })
    throw new ContainerException(concrete.getName + " does not have a satisfiable constructor", exceptions)
  }
}

File: src/com/googlecode/yadic/SimpleContainer.java
package com.googlecode.yadic;

import com.googlecode.yadic.resolvers.MissingResolver;
import com.googlecode.yadic.resolvers.Resolvers;

import java.lang.reflect.Type;
import java.util.concurrent.Callable;

import static com.googlecode.totallylazy.Functions.returns;
import static com.googlecode.yadic.resolvers.Resolvers.*;

public class SimpleContainer extends DelegatingTypeMap implements Container {
    private SimpleContainer(TypeMap typeMap) {
        super(typeMap);
    }

    public SimpleContainer(Resolver<?> parent) {
        this(new BaseTypeMap(parent));
    }

    public SimpleContainer() {
        this(new MissingResolver());
    }

    public static SimpleContainer container(TypeMap typeMap) {
        return new SimpleContainer(typeMap);
    }

    @Override
    protected TypeMap self() {
        return this;
    }

    @SuppressWarnings("unchecked")
    public <T> T get(Class<T> aClass) {
        return (T) Resolvers.resolve(this, aClass);
    }

    public <T> Callable<T> getActivator(Class<T> aClass) {
        return asCallable(super.<T>getResolver(aClass), aClass);
    }

    public <T> Container add(final Class<T> concrete) {
        addType((Type) concrete, concrete);
        return this;
    }

    public <I, C extends I> Container add(Class<I> anInterface, Class<C> concrete) {
        addType((Type) anInterface, concrete);
        return this;
    }

    public <I, C extends I> Container addInstance(Class<I> anInterface, final C instance) {
        return addActivator(anInterface, returns(instance));
    }

    @SuppressWarnings("unchecked")
    public <T, A extends Callable<T>> Container addActivator(Class<T> aClass, final Class<A> activator) {
        addType(aClass, activator(this, activator));
        return this;
    }

    public <T> Container addActivator(Class<T> aClass, final Callable<? extends T> activator) {
        addType(aClass, asResolver(activator));
        return this;
    }

    public <I, C extends I> Container decorate(final Class<I> anInterface, final Class<C> concrete) {
        addType(anInterface, decorator(this, anInterface, concrete));
        return this;
    }

    public <I, C extends I> Container replace(Class<I> anInterface, Class<C> newConcrete) {
        remove(anInterface);
        return add(anInterface, newConcrete);
    }

}